packages/flutter/lib/src/widgets/actions.dart

```diff
@@ -613,36 +618,68 @@ class _FocusableActionDetectorState extends State<FocusableActionDetector> {
   void _handleMouseEnter(PointerEnterEvent event) {
     assert(widget.onShowHoverHighlight != null);
     if (!_hovering) {
-      // TODO(gspencergoog): remove scheduleMicrotask once MouseRegion event timing has changed.
-      scheduleMicrotask(() { setState(() { _hovering = true; _handleShowHoverHighlight(); }); });
+      _mayTriggerCallback(task: () {
+        setState(() { _hovering = true; });
+      });
     }
   }
 
   void _handleMouseExit(PointerExitEvent event) {
     assert(widget.onShowHoverHighlight != null);
     if (_hovering) {
-      // TODO(gspencergoog): remove scheduleMicrotask once MouseRegion event timing has changed.
-      scheduleMicrotask(() { setState(() { _hovering = false; _handleShowHoverHighlight(); }); });
+      _mayTriggerCallback(task: () {
+        setState(() { _hovering = false; });
+      });
     }
   }
 
   bool _focused = false;
   void _handleFocusChange(bool focused) {
     if (_focused != focused) {
-      setState(() {
-        _focused = focused;
-        _handleShowFocusHighlight();
-        widget.onFocusChange?.call(_focused);
+      _mayTriggerCallback(task: () {
+        setState(() {
+          _focused = focused;
+          widget.onFocusChange?.call(_focused);
+        });
       });
     }
   }
 
-  void _handleShowHoverHighlight() {
-    widget.onShowHoverHighlight?.call(_hovering && widget.enabled && _canShowHighlight);
+  bool _shouldShowHoverHighlight({@required FocusableActionDetector target}) {
+    return _hovering && target.enabled && _canShowHighlight;
+  }
+  bool _shouldShowFocusHighlight({@required FocusableActionDetector target}) {
+    return _focused && target.enabled && _canShowHighlight;
+  }
+
+  // Record old states, do `task` if not null, then compare old states with the
+  // new state, and trigger callbacks if necessary.
+  //
+  // The old states are collected on `oldWidget` if it is provided, or the
+  // current widget (before doing `task`) otherwise.
+  void _mayTriggerCallback({VoidCallback task, FocusableActionDetector oldWidget}) {
```

Comment:
I think `task` here can just be a positional argument.

Reply:
The `_mayTriggerCallback` called in `didUpdateWidget` does not have a task, but an `oldWidget` instead.